toolchain/check/handle_loop_statement.cpp

```diff
@@ -75,20 +113,124 @@ auto HandleParseNode(Context& context, Parse::WhileStatementId node_id)
+  context.scope_stack().PushForSameRegion();
+
+  // Begin an implicit let declaration context for the pattern.
+  context.decl_introducer_state_stack().Push<Lex::TokenKind::Let>();
+  context.pattern_block_stack().Push();
+  context.full_pattern_stack().PushFullPattern(
+      FullPatternStack::Kind::NameBindingDecl);
+  BeginSubpattern(context);
+
+  context.node_stack().Push(node_id);
+  return true;
 }
 
 auto HandleParseNode(Context& context, Parse::ForInId node_id) -> bool {
-  context.decl_introducer_state_stack().Pop<Lex::TokenKind::Var>();
-  return context.TODO(node_id, "HandleForIn");
+  auto pattern_block_id = context.pattern_block_stack().Pop();
+  AddInst<SemIR::NameBindingDecl>(context, node_id,
+                                  {.pattern_block_id = pattern_block_id});
+  context.decl_introducer_state_stack().Pop<Lex::TokenKind::Let>();
+  context.full_pattern_stack().StartPatternInitializer();
+  context.node_stack().Push(node_id, pattern_block_id);
+  return true;
+}
+
+// For a value or reference of type `Optional(T)`, call the given accessor.
+static auto CallOptionalAccessor(Context& context, Parse::NodeId node_id,
+                                 SemIR::InstId optional_id,
+                                 llvm::StringLiteral accessor_name)
+    -> SemIR::InstId {
+  auto accessor_name_id =
+      SemIR::NameId::ForIdentifier(context.identifiers().Add(accessor_name));
+  auto accessor_id =
+      PerformMemberAccess(context, node_id, optional_id, accessor_name_id);
+  return PerformCall(context, node_id, accessor_id, {});
 }
 
 auto HandleParseNode(Context& context, Parse::ForHeaderId node_id) -> bool {
-  return context.TODO(node_id, "HandleForHeader");
+  auto range_id = context.node_stack().PopExpr();
+  auto pattern_block_id = context.node_stack().Pop<Parse::NodeKind::ForIn>();
+  auto pattern_id = context.node_stack().PopPattern();
+  auto start_node_id =
+      context.node_stack().PopForSoloNodeId<Parse::NodeKind::ForHeaderStart>();
+
+  // Convert the range expression to a value or reference so that we can use it
+  // multiple times.
+  // TODO: If this produces a temporary, its lifetime should presumably be
+  // extended to cover the loop body.
+  range_id = ConvertToValueOrRefExpr(context, range_id);
+
+  // Create the cursor variable.
+  // TODO: Produce a custom diagnostic if the range operand can't be used as a
+  // range.
+  auto cursor_id = BuildUnaryOperator(
+      context, node_id, {.interface_name = "Iterate", .op_name = "NewCursor"},
+      range_id);
+  auto cursor_type_id = context.insts().Get(cursor_id).type_id();
+  auto cursor_var_id = AddInstWithCleanup(
+      context, node_id,
+      SemIR::VarStorage{.type_id = cursor_type_id,
+                        .pattern_id = SemIR::AbsoluteInstId::None});
+  auto init_id = Initialize(context, node_id, cursor_var_id, cursor_id);
+  AddInst<SemIR::Assign>(context, node_id,
+                         {.lhs_id = cursor_var_id, .rhs_id = init_id});
```

Comment:
I don't really care about type placement in `AddInst<SemIR::Assign>` versus `AddInstWithCleanup(... SemIR::VarStorage...);`, but you might consider aligning to use the same approach in both spots. Here's one suggestion. :)

```suggestion
  auto cursor_var_id = AddInstWithCleanup(
      context, node_id,
      SemIR::VarStorage{.type_id = cursor_type_id,
                        .pattern_id = SemIR::AbsoluteInstId::None});
  auto init_id = Initialize(context, node_id, cursor_var_id, cursor_id);
  AddInst(context, node_id,
          SemIR::Assign{.lhs_id = cursor_var_id, .rhs_id = init_id});
```

Reply:
Done. I went in the opposite direction given that `AddInst<T>(..., {...})` is much more common than `AddInst(..., T{...})` in existing code.